### ISSUE TITLE
Custom implementation of unification

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#1661]
Right now we use the `unification-fd` package to implement [unification](https://en.wikipedia.org/wiki/Unification_(computer_science)) of types during type checking, [as described here](https://byorgey.wordpress.com/2021/09/08/implementing-hindley-milner-with-the-unification-fd-library/).  However, I think perhaps it's time to replace our use of `unification-fd` with our own custom implementation of unification.  This would have a number of benefits:

- `unification-fd` is quite old and it has some idiosyncracies that have been awkward to work around.  For example, it insists on working in the context of a monad transformer stack that has an error transformer on top of its own `IntBindingT` transformer.  If we implement our own we can integrate it with our existing effects framework instead.
- Implementing our own unification means the possibility to generate better custom error messages.
- `unification-fd` can only do pure structural unification.  However, several features we are contemplating might need more custom processing:
    - Type synonyms (#153) would require expanding type synonyms to their definitions whenever they are encountered during unification
    - Recursive types (#154) require being able to unroll recursive types one step when they are encountered during unification.
        - With equirecursive types, we also would have to track a set of encountered pairs during unification so that we can cut off the recursion when seeing a given pair again
        - Even with isorecursive types, it is too fragile to require that a recursive type be available in the right mode (check or infer) when encountering every `roll` or `unroll`; we need to be able to solve for unification variables matched with recursive types, which means being able to unroll them as needed during unification.

On the other hand there would also be some downsides:

- `unification-fd` is quite sophisticated and optimized; it is almost surely way faster than anything we might implement ourselves.  However, I am hoping that this will not make too much of a difference as long as our new implementation is "good enough".  And even if it isn't, we might be able to incorporate some optimization techniques from `unification-fd` into our own implementation.
- It represents a significant refactoring effort.

**Comment by byorgey** [comment]
Some potentially helpful resources:
- https://en.wikipedia.org/wiki/Unification_(computer_science)
- https://github.com/YueLiPicasso/unification
- http://strictlypositive.org/foubsr-website/

**Comment by byorgey** [comment]
It may look like I haven't been very active recently --- and it's I haven't had as much time to work on Swarm in general --- but in fact I've been working steadily on this.  In the [`refactor/custom-unification` branch](https://github.com/swarm-game/swarm/tree/refactor/custom-unification) I have finally succeeded in completely ripping out `unification-fd` and replacing it with a custom (naive) implementation which passes all the tests.  Unfortunately it is quite a bit slower (running the entire test suite went from around 120s on my machine to 190s, a 60% increase).  I am currently reading [Efficient Functional Unification and Substitution](http://webdoc.sub.gwdg.de/ebook/serien/ah/UU-CS/2008-027.pdf) and hopeful it may provide ideas for an implementation which is fast enough but not terribly complex (I have looked into the guts of `unification-fd` and it is so highly optimized and complex that I despair of understanding it).

Once I finish this I will be able to tackle #154 which I am very excited about.  Then eventually I'll get back around to #1087 .
